### PR TITLE
dataflow: add a shutdown button for kafka sinks

### PR DESCRIPTION
**Pull request by ruchirK**
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/3353)

### CODE REVIEW

**Comment by ruchirK** [comment]
I think something might be wrong in my changes - when I add this diff
```
altaria-2:billing Test$ git diff
diff --git a/src/dataflow/src/server.rs b/src/dataflow/src/server.rs
index f5daf1cb..2df586a7 100644
--- a/src/dataflow/src/server.rs
+++ b/src/dataflow/src/server.rs
@@ -538,6 +538,7 @@ where
             }
             SequencedCommand::DropSinks(ids) => {
                 for id in ids {
+                    log::debug!("dropping sink {}", id);
                     self.sink_tokens.remove(&id);
                 }
             }
diff --git a/src/dataflow/src/sink/kafka.rs b/src/dataflow/src/sink/kafka.rs
index d0a4bdd5..4f3ff32e 100644
--- a/src/dataflow/src/sink/kafka.rs
+++ b/src/dataflow/src/sink/kafka.rs
@@ -15,7 +15,7 @@ use std::time::Duration;
 use differential_dataflow::hashable::Hashable;
 use differential_dataflow::operators::arrange::ShutdownButton;
 use lazy_static::lazy_static;
-use log::error;
+use log::{debug, error};
 use prometheus::{
     register_int_counter_vec, register_uint_gauge_vec, IntCounter, IntCounterVec, UIntGauge,
     UIntGaugeVec,
@@ -294,6 +294,7 @@ where

                         false
                     } else {
+                        debug!("producer None for {}", name);
                         false
                     }
                 },
altaria-2:billing Test$
```

and then drop a sink I never observe a log line saying `producer None ...`. In addition looking at metrics it seems like the sink always runs to completion even after its dropped

File: src/dataflow/src/sink/util.rs [hunk 1]
```diff
@@ -7,28 +7,36 @@
...
     D: Data,
-    B: FnOnce(OperatorInfo) -> L,
+    B: FnOnce(OperatorInfo) -> (L, ShutdownButton<T>),
     L: FnMut(&mut FrontieredInputHandle<G::Timestamp, D, P::Puller>) -> bool + 'static,
     P: ParallelizationContract<G::Timestamp, D>,
 {
     let mut builder = OperatorBuilder::new(name.to_owned(), s.scope());
     let operator_info = builder.operator_info();
     let mut input = builder.new_input(s, pact);
+    let (mut logic, button) = constructor(operator_info);
```
Comment (ruchirK) [inline src/dataflow/src/sink/util.rs:32]: note: I don't fully understand the consequences of lifting this function call out of the closure

**Comment by ruchirK** [comment]
I think I have the fix -- wasn't calling `press_on_drop()` (which I swear I had already :/) will report back if it fixes

**Comment by ruchirK** [comment]
Fixed this! Ok everything looks good and the metrics all indicate that we are doing the right stuff :)

File: src/dataflow/src/sink/kafka.rs [hunk 2]
```diff
@@ -176,105 +183,122 @@ pub fn kafka<G>(
...
-                // send it over to Kafka. We Even though we want to do bounded work
-                // per sink invocation, we still need to remember all inputs as we
-                // receive them
-                input.for_each(|_, rows| {
-                    rows.swap(&mut vector);
+            let shutdown_button_ref = &mut shutdown_button;
+            *shutdown_button_ref = Some(ShutdownButton::new(
+                producer.clone(),
+                stream.scope().activator_for(&info.address[..]),
+            ));
```
Comment (benesch) [inline src/dataflow/src/sink/kafka.rs:190]: I'm not sure I understand what this is all about. Why can't you ditch the option like this?

```rust
            let shutdown_button = ShutdownButton::new(
                producer.clone(),
                stream.scope().activator_for(&info.address[..]),
            );
```

File: src/dataflow/src/sink/util.rs (hunk 1 above)
Comment (benesch) [reply src/dataflow/src/sink/util.rs:32]: AFAICT it would only matter if `constructor` needed access to `capabilities`, but it does not.

File: src/dataflow/src/sink/kafka.rs [hunk 3]
```diff
@@ -176,105 +183,122 @@ pub fn kafka<G>(
...
 
-                    for (row, _time, diff) in vector.drain(..) {
-                        queue.push_back((row, diff));
-                    }
-                });
+            (
+                move |input| {
+                    let producer = producer.borrow();
+                    if let Some(ref producer) = *producer {
```
Comment (benesch) [inline src/dataflow/src/sink/kafka.rs:195]: FWIW I consider this alternative formulation to be pretty idiomatic:

```rust
if let Some(producer) = &*producer.borrow() {
```

File: src/dataflow/src/sink/kafka.rs [hunk 4]
```diff
@@ -176,105 +183,122 @@ pub fn kafka<G>(
...
+                            // Let's make sure to keep the sink operator around until
+                            // we flush them out
+                            activator.activate_after(Duration::from_secs(5));
+                            return true;
+                        }
 
-                false
-            }
+                        false
+                    } else {
```
Comment (benesch) [inline src/dataflow/src/sink/kafka.rs:296]: The level of nesting in this function has made things pretty hard to follow. Probably time to extract into a helper function?

File: src/dataflow/src/sink/kafka.rs (hunk 2 above)
Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs]: I was copying the pattern that was used for creating shutdown buttons in differential but agreed it doesnt seem necessary in this case so I've removed. thanks for the catch!

File: src/dataflow/src/sink/kafka.rs (hunk 4 above)
Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs]: I've tried to change it to check for is_none and return early in that case, which avoids the extra level of nesting. let me know if that feels good enough.

I'm happy to extract into a helper instead but it seems kind of hard because the innermost closure uses so many variables defined in outer scopes. If you feel strongly about this though please lmk

File: src/dataflow/src/sink/kafka.rs (hunk 3 above)
Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs]: I've seen this reformulation a bunch of times and I am not sure I grok it - to check my understanding:
is the right way to read this from right to left

1. the `borrow()` gives us a reference to the object inside the refcell
2. the `*` dereferences that reference (giving us an option)
3. the `&` takes a reference to whats inside the option? or does it just take a reference to the option again?

the last bit is the most confusing one

File: src/dataflow/src/sink/kafka.rs (hunk 4 above)
Comment (ruchirK) [reply src/dataflow/src/sink/kafka.rs]: I also changed it to bind the inner closure to a variable and then return the pair (closure, shutdownbutton) separately which also reduces a level of nesting

File: src/dataflow/src/sink/kafka.rs [hunk 5]
```diff
@@ -176,7 +184,21 @@ pub fn kafka<G>(
...
+
+            let ret = move |input: &mut FrontieredInputHandle<_, _, _>| {
+                let producer: &RefCell<_> = producer.borrow();
+                let producer = &*producer.borrow();
+                if producer.is_none() {
+                    // The sink has already been dropped
+                    return false;
+                }
+
+                let producer = producer.as_ref().expect("producer known to exist");
```
Comment (benesch) [inline src/dataflow/src/sink/kafka.rs:200]: Btw, the Rusty pattern for pattern-matching early returns is to use a match statement with the return inside:

```rust
let producer = match &*producer.borrow() {
    Some(producer) => producer,
    None => return false,
};
```
